Question 1: The Road Trip Planner
Given:
Fuel efficiency = 11.5 km/l


Total trip distance = 475 km


Fuel price = $1.65/l


Tank capacity = 50 liters


You start with a full tank


In [4]:
#a
d = 475  # in km
f_efficiency = 11.5  # in km per liter

f_needed = d / f_efficiency
print(f"Fuel needed: {f_needed:.2f} liters")

Fuel needed: 41.30 liters


a. I divide the total distance by the car's fuel efficiency to calculate the total fuel required for trip.
475 / 11.5 = 41.30 liters. So, the car will need 41.3 liters of fuel to cover 475 kilometers.


In [5]:
#b
t_capacity = 50  # in liters

full_tanks = f_needed // t_capacity
print(f"Full tanks required: {int(full_tanks)}")

Full tanks required: 0


b. I used the floor division operator // to find how many whole full tanks are required.
41.3 // 50 = 0.0 → This means less than one full tank is needed. So, only part of the full tank will be used.


In [6]:
#c
f_price = 1.65  #  per liter

total_cost = f_needed * f_price
print(f"Total fuel cost: ${total_cost:.2f}")

Total fuel cost: $68.15


c. To get the cost, multiply the amount of fuel needed by the price per liter. 41.3 * 1.65 = $68.145, rounded to $68.15


In [7]:
#d
fuel_remaining = t_capacity - f_needed
print(f"Fuel remaining after trip: {fuel_remaining:.2f} liters")

Fuel remaining after trip: 8.70 liters


d. If I started with a full tank (50 liters) and only needed 41.3 liters for
the trip, the fuel left after the  trip would be: 50 - 41.3 = 8.7 liters.


---




Question 2: Theme Park Entry Rules
A person gets a discount if any ONE of these is true:

Child (≤12 years old) AND it's a weekday

Senior (≥65 years old)

It's Family Day AND group size ≥ 4

In [8]:
#a
age = 30
is_weekday = True
is_family_day = False
group_size = 3

discount = (age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)
discount

False

a. No condition is satisfied with the given conditions.

In [9]:
#b
age = 10
is_weekday = True
is_family_day = False
group_size = 2

discount = (age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)

b. Age ≤ 12 and it's a weekday.

c.
age <= 12 → True

is_weekday → True

First part: True and True → True

Rest not checked due to or short-circuit

Final result: True


---



Question 3: Message Decoder
You are given the following secret message encoded in a Python string:
secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"


In [10]:
secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"


In [11]:
#a Index 5
secret_message[5]

'N'

a. As the index starts from zero the 5th index is 'N'

In [12]:
#b Slice "Is_FuN"
secret_message[7:13]

'Is_FuN'

In [13]:
#c Slice "PoWeRfUl" using negative indexing
secret_message[-9:-1]

'PoWeRfUl'

In [22]:
#d Every 4th character
secret_message[::4]

'PosNDWU'

In [14]:
#e
# secret_message[50]

e. Index 50 is beyond the string's length, so Python throws an IndexError.



---



Section 4: Boolean Values & Expressions
(x == z) and (y > x) = True,
(x != z) or (y < x) or (len(text) < 5) = False,
not ((x > y) and (text == "Adventure")) = True

In [15]:
x = 15
y = 25
z = 15
text = "adventure"
print((x == z) and (y > x))
print((x != z) or (y < x) or (len(text) < 5))
print(not ((x > y) and (text == "Adventure")))

True
False
True


EXPLANATION:
True and True = True,
False or False or False = False,
not (False and False) = not False = True

> Add blockquote



Section 5: String Properties & Methods


In [ ]:
#a
username = "MyUser123"
username[0] = 'm'

a. strings are immutable

In [16]:
#b.
quote = " practice makes perfect "
formatted_quote = quote.strip().upper().replace("PERFECT", "PROGRESS")
formatted_quote

'PRACTICE MAKES PROGRESS'

strip() removes spaces,
upper() converts to uppercase,
replace() swaps PERFECT for PROGRESS

In [17]:
#c.
len_original = len(quote)          # 25
len_formatted = len(formatted_quote)  # 24
print(len_original, len_formatted)

24 23
